x-pack/filebeat/input/entityanalytics/provider/activedirectory/activedirectory.go

```diff
@@ -210,48 +214,86 @@ func (p *adInput) runFullSync(inputCtx v2.Context, store *kvstore.Store, client
 
 	ctx := ctxtool.FromCanceller(inputCtx.Cancelation)
 	p.logger.Debugf("Starting fetch...")
-	_, err = p.doFetchUsers(ctx, state, true)
+	users, err := p.doFetchUsers(ctx, state, true)
 	if err != nil {
-		return err
+		return time.Time{}, err
 	}
 
-	if len(state.users) != 0 {
-		tracker := kvstore.NewTxTracker(ctx)
-
-		start := time.Now()
-		p.publishMarker(start, start, inputCtx.ID, true, client, tracker)
-		for _, u := range state.users {
-			p.publishUser(u, state, inputCtx.ID, client, tracker)
+	var (
+		tracker *kvstore.TxTracker
+		latest  time.Time
+	)
+	if len(users) != 0 || state.len() != 0 {
+		// Active Directory does not have a notion of deleted users
+		// beyond absence from the directory, so compare found users
+		// with users already known by the state store and if any
+		// are in the store but not returned in the previous fetch,
+		// mark them as deleted and publish the deletion. We do not
+		// have the time of the deletion, so use now.
+		if state.len() != 0 {
+			found := make(map[string]bool)
+			for _, u := range users {
+				found[u.ID] = true
+			}
+			deleted := make(map[string]*User)
+			now := time.Now()
+			state.forEach(func(u *User) {
+				if u.State == Deleted {
+					// We have already seen that this is deleted
+					// so we do not need to publish again. The
+					// user will be deleted from the store when
+					// the state is closed.
+					return
+				}
+				if found[u.ID] {
+					// We have the user, so we do not need to
+					// mark it as deleted.
+					return
+				}
+				// This modifies the state store's copy since u
+				// is a pointer held by the state store map.
+				u.State = Deleted
+				u.WhenChanged = now
+				deleted[u.ID] = u
+			})
+			for _, u := range deleted {
+				users = append(users, u)
+			}
+		}
+		if len(users) != 0 {
+			tracker = kvstore.NewTxTracker(ctx)
+			for _, u := range users {
+				p.publishUser(u, state, inputCtx.ID, client, tracker)
+				if u.WhenChanged.After(latest) {
```

Comment:
Also, `doFetchUsers` updates `state.WhenChanged`, so that could be read back, and it would only have the times from the server.

Reply:
I'm torn on this last point. I see the merit of the efficiency improvement, and did consider it when I was writing this, but I really don't like the massive amount of data flow by shared pointers. The lack of clarity it brings is significant. I'll do this, but it needs good commentary.

Reply:
I think either way is okay. I'm only thinking about the clarity.

Adding an "and not deleted" condition is good because it's close to where the deletions are made, so probably clear enough.

The intent is for `latest` to be set to the last time we saw from the server. After we call something that updates `state` with information from the server, reading it from there seems clearer than recalculating it.

On the other hand, there's probably some clean-up necessary to be really clear about where different responsibilities lie.

Reply:
I believe this is done now.